File: Books/A_Comprehensive_Approach_to_ML_and_AI/matlab/svd_tutorial.m
% Example – PCA to compress color images
%  - Recall that a color image is made up of pixels with red, green, blue 
%    (RGB) values.
%  - Essentially, we have a collection of 3D vectors in RGB space.


% Start of script
%-------------------------------------------------------------------------%
close all;                   	% close all figures
clearvars; clearvars -global;	% clear all variables
clc;                         	% clear the command terminal
format shortG;                 	% pick the most compact numeric display
format compact;                	% suppress excess blank lines

% Read data & plot
%-------------------------------------------------------------------------%
RGB = im2double(imread('data/arizona_photo.jpg'));

% Convert 3-dimensional array array to 2D, where each row is a pixel (RGB)
X = reshape(RGB, [], 3);
N = size(X,1); % N is the number of pixels

% Plot pixels in color space. To limit the number of points, only plot
% every 100th point.
figure
hold on
for i=1:100:size(X,1)
    mycolor = X(i,:);
    mycolor = max(mycolor, [0 0 0]);
    mycolor = min(mycolor, [1 1 1]);
    plot3(X(i, 1), X(i, 2), X(i, 3),'.', 'Color', mycolor);
end
xlabel('red'), ylabel('green'), zlabel('blue');
xlim([0 1]), ylim([0 1]), zlim([0 1]);
hold off
axis equal

% Do PCA on color vectors ... keep the top two PCs.
%-------------------------------------------------------------------------%
% Get mean and covariance
m_x = mean(X);
C_x = cov(X);

% Get eigenvalues and eigenvectors of C_x.
% Produces V,D such that C_x*V = V*D.
% So the eigenvectors are the columns of V.
[V,D] = eig(C_x);
e1 = V(:,3);
disp('Eigenvector e1:'), disp(e1);
e2 = V(:,2);
disp('Eigenvector e2:'), disp(e2);
e3 = V(:,1);
disp('Eigenvector e3:'), disp(e3);
d1 = D(3,3);
disp('Eigenvalue d1:'), disp(d1);
d2 = D(2,2);
disp('Eigenvalue d2:'), disp(d2);
d3 = D(1,1);
disp('Eigenvalue d3:'), disp(d3);

% Construct matrix A such that the 1st row of A is the eigenvector 
% corresponding to the largest eigenvalue, the 2nd row is the eigenvector 
% corresponding to the second largest eigenvalue, etc. 
A = [e1'; e2'; e3'];

% Project input vectors x onto eigenvectors. For each (column) vector x,
% we will use the equation y = A*(x - mx).
% To explain the Matlab commands below:
% X is our (N,3) array of vectors; each row is a vector.
% mx is the mean of the vectors, size (1,3).
% We first subtract off the mean using X - repmat(mx,N,1).
% We then transpose that result so that each vector is a column.
% We then apply our transform A to each column.
Y = A*(X - repmat(m_x,N,1))'; % Y has size 3xN

% Display y vectors as images 
[height,width,depth] = size(RGB); 
Y1 = reshape(Y(1,:), height, width); 
Y2 = reshape(Y(2,:), height, width); 
Y3 = reshape(Y(3,:), height, width); 
figure; 
subplot(1,3,1), imshow(Y1,[]); 
subplot(1,3,2), imshow(Y2,[]); 
subplot(1,3,3), imshow(Y3,[]);

% Reconstruct image using only Y1 and Y2. For each (column) vector y,
% we will use the equation x = A'*y + mx.
% To explain the Matlab commands below:
% Y is our (3,N) array of vectors; where each column is a vector.
% A(1:k,:) is the first k rows of A.
% Y(1:k,:) is the first k rows of Y.
% A(1:k,:)' * Y(1:k,:) produces our transformed vectors (3xN); we then
% transpose that to make an array of size Nx3, and add the mean.
k = 1;
A_k = A(1:k,:);
X_r = (A(1:k,:)' * Y(1:k,:))' + repmat(m_x,N,1);

figure
hold on
for i=1:100:size(X_r,1)
    mycolor = X_r(i,:);
    mycolor = max(mycolor, [0 0 0]);
    mycolor = min(mycolor, [1 1 1]);
    plot3(X_r(i, 1), X_r(i, 2), X_r(i, 3),'.', 'Color', mycolor);
end
xlabel('red'), ylabel('green'), zlabel('blue');
xlim([0 1]), ylim([0 1]), zlim([0 1]);
hold off
axis equal

% Display original image and reconstructed image
figure, subplot(1,2,1), imshow(RGB), title('Original');
Ir(:,:,1) = reshape(X_r(:,1), height, width);
Ir(:,:,2) = reshape(X_r(:,2), height, width);
Ir(:,:,3) = reshape(X_r(:,3), height, width);
subplot(1,2,2), imshow(Ir), title('Reconstructed');

% Display original image and reconstructed image
figure
imshow(RGB)
title('Original');

Ir(:,:,1) = reshape(X_r(:,1), height, width);
Ir(:,:,2) = reshape(X_r(:,2), height, width);
Ir(:,:,3) = reshape(X_r(:,3), height, width);

figure
imshow(Ir)
title('Reconstructed');

% save_all_figs_OPTION('results/svd','png',0)

File: Books/A_Comprehensive_Approach_to_ML_and_AI/python/svd_tutorial.py
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from skimage import io

# 1. Read the image (ensure the image file is in the 'data' folder)
RGB = io.imread('data/arizona_photo.jpg') / 255.0  # convert to [0,1] float

# 2. Reshape the image: each row is a pixel [R, G, B]
height, width, channels = RGB.shape
X = RGB.reshape(-1, 3)  # shape (N, 3)
N = X.shape[0]

# 3. Plot pixels in RGB space (plot every 100th pixel)
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
for i in range(0, N, 100):
    ax.scatter(X[i, 0], X[i, 1], X[i, 2], color=X[i, :], s=20)
ax.set_xlabel('Red')
ax.set_ylabel('Green')
ax.set_zlabel('Blue')
ax.set_xlim([0,1]); ax.set_ylim([0,1]); ax.set_zlim([0,1])
plt.title('RGB Color Space Scatter Plot')
plt.show()

# 4. Perform PCA on color vectors
m_x = np.mean(X, axis=0)
X_centered = X - m_x
C_x = np.cov(X_centered, rowvar=False)

# Compute eigenvalues and eigenvectors.
eigvals, eigvecs = np.linalg.eig(C_x)
# Sort eigenvalues in ascending order and reorder eigenvectors.
sorted_idx = np.argsort(eigvals)
e1 = eigvecs[:, sorted_idx[-1]]
e2 = eigvecs[:, sorted_idx[-2]]
e3 = eigvecs[:, sorted_idx[-3]]
d1 = eigvals[sorted_idx[-1]]
d2 = eigvals[sorted_idx[-2]]
d3 = eigvals[sorted_idx[-3]]
print("Eigenvector e1:", e1)
print("Eigenvector e2:", e2)
print("Eigenvector e3:", e3)
print("Eigenvalue d1:", d1)
print("Eigenvalue d2:", d2)
print("Eigenvalue d3:", d3)

# Construct transformation matrix A: rows are eigenvectors ordered by descending eigenvalue.
A = np.vstack([e1, e2, e3])

# 5. Project pixels: Y = A * (x - m_x) for each pixel.
Y = (A @ X_centered.T)  # shape (3, N)

# Display the projections as images.
Y1 = Y[0, :].reshape(height, width)
Y2 = Y[1, :].reshape(height, width)
Y3 = Y[2, :].reshape(height, width)
plt.figure(figsize=(15,4))
plt.subplot(1,3,1); plt.imshow(Y1, cmap='gray'); plt.title('Projection on PC1')
plt.subplot(1,3,2); plt.imshow(Y2, cmap='gray'); plt.title('Projection on PC2')
plt.subplot(1,3,3); plt.imshow(Y3, cmap='gray'); plt.title('Projection on PC3')
plt.show()

# 6. Reconstruct image using only the first principal component.
k = 1
A_k = A[:k, :]  # (1, 3)
# Project and reconstruct: X_r = (A_k^T * Y_k) + m_x, where Y_k = A_k*(X - m_x)
Y_k = A_k @ X_centered.T   # shape (1, N)
X_r = (A_k.T @ Y_k).T + m_x  # shape (N, 3)
X_r = np.clip(X_r, 0, 1)
# Reshape back to image.
reconstructed = X_r.reshape(height, width, 3)

# Plot color-space scatter of reconstructed pixels.
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
for i in range(0, N, 100):
    ax.scatter(X_r[i, 0], X_r[i, 1], X_r[i, 2], color=X_r[i, :], s=20)
ax.set_xlabel('Red')
ax.set_ylabel('Green')
ax.set_zlabel('Blue')
ax.set_xlim([0,1]); ax.set_ylim([0,1]); ax.set_zlim([0,1])
plt.title('Reconstructed Color Space Scatter (using PC1)')
plt.show()

# Display original and reconstructed images.
plt.figure()
plt.subplot(1,2,1); plt.imshow(RGB); plt.title('Original Image')
plt.subplot(1,2,2); plt.imshow(reconstructed); plt.title('Reconstructed Image (PC1)')
plt.show()
